File: Common/Src/ChinesePuzzle/GameScene.cpp
/**
 *  GameScene.cpp
 *  ChinesePuzzle
 *
 *  Created by Mathieu LEDRU on 01/11/11.
 *
 *  GPL License:
 *  Copyright (c) 2011, Mathieu LEDRU
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include "GameScene.h"
#include "SimpleAudioEngine.h"

using namespace cocos2d;
using namespace CocosDenshion;

typedef enum {
	BGMusicThemeNone,
	BGMusicTheme1,
	BGMusicTheme2,
	BGMusicTheme3,
} BGMusicTheme;

GameSceneCommon::GameSceneCommon() :
conf(NULL),
bgMusicTheme(BGMusicThemeNone)
{
	
}

GameSceneCommon::~GameSceneCommon()
{
	CC_SAFE_RELEASE(conf);
	CC_SAFE_RELEASE(pGame);
	CC_SAFE_RELEASE(pMenu);
}

bool GameSceneCommon::init()
{
	if (!CCScene::init())
	{
		return false;
	}
	
	this->conf = new GameConfig();
	this->conf->init();
	this->conf->load();
	
	playBackgroundMusic(this->conf->getIsSoundOn());
	
	pGame = NULL;
	pMenu = NULL;
	
	this->game();
	
	return true;
}

Game* GameSceneCommon::game()
{
	if(pMenu)
	{
		//remove menu when getting the game
		this->removeChild(pMenu, true);
		CC_SAFE_RELEASE_NULL(pMenu);
	}
	
	if (pGame == NULL)
	{
		pGame = Game::node(this);
		pGame->retain();
	}
	
	if(pGame->getParent() == NULL)
	{
		this->addChild(pGame, GameSceneZOrderGame);
	}
	
	return pGame;
}

Menu* GameSceneCommon::menu()
{
	return this->menuWithLayout(MenuLayout::TypeNone);
}

Menu* GameSceneCommon::menuWithLayout(MenuLayout::Type ml)
{
	if (pMenu == NULL)
	{
		MenuInit mi;
		mi.gs = this;
		mi.layout = ml;
		
		pMenu = Menu::node(mi);
		pMenu->retain();
	}
	
	if(pMenu->getParent() == NULL)
	{
		this->addChild(pMenu, GameSceneZOrderMenu);
	}
	
	return pMenu;
}

void GameSceneCommon::newGame()
{
	Game* aGame = this->game();
	aGame->newGame();
	
	playSound("shuffle");
}

void GameSceneCommon::retryGame()
{
	Game* aGame = this->game();
	aGame->retryGame();
	
	playSound("shuffle");
}

void GameSceneCommon::setResolution(const std::string& resolution)
{
	conf->setResolution(resolution);
	this->setContentSize(conf->getResolutionSize());
	
	this->layout(false);
	conf->save();
}

void GameSceneCommon::setTheme(const std::string& theme)
{
	conf->setTheme(theme);
	this->layout();
	
	conf->save();
}

void GameSceneCommon::playSound(const char* soundName)
{
	SimpleAudioEngine* audio = SimpleAudioEngine::sharedEngine();
	
	if(conf->getIsSoundOn())
	{
		std::string soundPath = "sound/";
		audio->playEffect((soundPath + soundName + ".caf").c_str());
	}
}

void GameSceneCommon::playBackgroundMusic(bool play)
{
	SimpleAudioEngine* audio = SimpleAudioEngine::sharedEngine();
	
	if(play && conf->getIsSoundOn())
	{
		audio->setBackgroundMusicVolume(0.5f);
		switch(bgMusicTheme)
		{
			case BGMusicThemeNone:
			case BGMusicTheme3: audio->playBackgroundMusic("sound/bgm1.caf", true); bgMusicTheme = BGMusicTheme1; break;
			case BGMusicTheme1: audio->playBackgroundMusic("sound/bgm2.caf", true); bgMusicTheme = BGMusicTheme2; break;
			case BGMusicTheme2: audio->playBackgroundMusic("sound/bgm3.caf", true); bgMusicTheme = BGMusicTheme3; break;
		}
	}
	else if(bgMusicTheme != BGMusicThemeNone)
	{
		audio->stopBackgroundMusic();
	}
}

void GameSceneCommon::layout(bool anim)
{
	if(pGame) pGame->layout(anim);
	if(pMenu) pMenu->layout(anim);
}

File: Common/Bin/Data/js/GameScene.js
/*
This file is part of the ChinesePuzzle package.

(c) Mathieu Ledru

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.
*/

cpz.GameSceneZOrder = {
  BG: 0,
  Game: 1,
  Menu: 2
};

cpz.GameSceneBGMusicTheme = {
  ThemeNone: 0,
  Theme1: 1,
  Theme2: 2,
  Theme3: 3
};

cpz.GameSceneCommon = cc.Scene.extend({
  _bgMusicTheme: null,
  _conf: null,
  _game: null,
  _menu: null,
  ctor: function() {
    this._super();
    return this._bgMusicTheme = cpz.GameSceneBGMusicTheme.ThemeNone;
  },
  init: function() {
    var lang;
    if (!this._super()) {
      return false;
    }
    lang = cc.CCLang.getInstance();
    lang.setLang(cc.LANGUAGE_FRENCH);
    lang.addLang('lang');
    this._conf = new cpz.GameConfig();
    this._conf.init();
    this._conf.load();
    this.playBackgroundMusic(this._conf.getIsSoundOn());
    this._game = null;
    this._menu = null;
    this.game();
    return true;
  },
  game: function() {
    if (this._menu) {
      this.removeChild(this._menu, true);
      this._menu = null;
    }
    if (this._game === null) {
      this._game = cpz.Game.create(this);
    }
    if (this._game.getParent() === null) {
      this.addChild(this._game, cpz.GameSceneZOrder.Game);
    }
    return this._game;
  },
  menu: function() {
    return this.menuWithLayout(cpz.MenuLayoutType.None);
  },
  menuWithLayout: function(layout) {
    if (this._menu === null) {
      this._menu = cpz.Menu.create(this, layout);
    }
    if (this._menu.getParent() === null) {
      this.addChild(this._menu, cpz.GameSceneZOrder.Menu);
    }
    return this._menu;
  },
  newGame: function() {
    var game;
    game = this.game();
    game.newGame();
    this.playSound('shuffle');
    return this;
  },
  retryGame: function() {
    var game;
    game = this.game();
    game.retryGame();
    this.playSound('shuffle');
    return this;
  },
  setResolution: function(resolution) {
    this._conf.setResolution(resolution);
    this.setContentSize(this._conf.getResolutionSize());
    this.layout(false);
    this._conf.save();
    return this;
  },
  setTheme: function(theme) {
    this._conf.setTheme(theme);
    this.layout();
    this._conf.save();
    return this;
  },
  playSound: function(soundName) {
    var audio;
    audio = cc.AudioEngine.getInstance();
    if (this._conf.getIsSoundOn()) {
      audio.playEffect(cpz.CommonPath + 'sound/' + soundName + '.mp3');
    }
    return this;
  },
  playBackgroundMusic: function(play) {
    var audio;
    audio = cc.AudioEngine.getInstance();
    if (play && this._conf.getIsSoundOn()) {
      audio.setMusicVolume(0.5);
      switch (this._bgMusicTheme) {
        case cpz.GameSceneBGMusicTheme.Theme1:
          audio.playMusic(cpz.CommonPath + 'sound/bgm2.mp3', true);
          return this._bgMusicTheme = cpz.GameSceneBGMusicTheme.Theme2;
        case cpz.GameSceneBGMusicTheme.Theme2:
          audio.playMusic(cpz.CommonPath + 'sound/bgm3.mp3', true);
          return this._bgMusicTheme = cpz.GameSceneBGMusicTheme.Theme3;
        default:
          audio.playMusic(cpz.CommonPath + 'sound/bgm1.mp3', true);
          return this._bgMusicTheme = cpz.GameSceneBGMusicTheme.Theme1;
      }
    } else if (this._bgMusicTheme !== cpz.GameSceneBGMusicTheme.ThemeNone) {
      return audio.stopMusic();
    }
  },
  layout: function(anim) {
    if (anim == null) {
      anim = true;
    }
    if (this._game) {
      this._game.layout(anim);
    }
    if (this._menu) {
      this._menu.layout(anim);
    }
    return this;
  },
  getConf: function() {
    return this._conf;
  },
  getGame: function() {
    return this._game;
  },
  getMenu: function() {
    return this._menu;
  }
});

cpz.GameSceneCommon.create = function() {
  var obj;
  obj = new cpz.GameScene();
  if (obj && obj.init()) {
    return obj;
  }
  return null;
};
